src/printer/tvmscript_unified_printer.cc

```diff
@@ -0,0 +1,738 @@
+  };
+
+  virtual void PrintDoc(const FunctionArgDoc& doc) override {
+    PrintDoc(doc->name);
+    output_ << ": ";
+    PrintDoc(doc->type);
+  };
+
+ private:
+  String tir_prefix_;
+
+  int indent_spaces_ = 4;
+
+  void IncreaseIndent() { indent_ += indent_spaces_; }
+
+  void DecreaseIndent() { indent_ -= indent_spaces_; }
+
+  void PrintWithIncreasedIndent(const Doc& doc) {
+    IncreaseIndent();
+    PrintDoc(doc);
+    DecreaseIndent();
+  }
+
+  template <typename ContainerType>
+  void PrintJoinedElements(const std::string& left, const ContainerType& elements,
+                           const std::string& separator, const std::string& right,
+                           const std::string& default_content = "") {
+    output_ << left;
+    bool is_first = true;
+    for (auto& element : elements) {
+      if (is_first) {
+        is_first = false;
+      } else {
+        output_ << separator;
+      }
+      PrintObject(element);
+    }
+    if (is_first) {
+      // This means that `elements` is empty
+      output_ << default_content;
+    }
+    output_ << right;
+  }
+
+  template <typename... Arg>
+  void PrintTIRPrimitiveCall(const std::string& primitive, Arg... args) {
+    output_ << tir_prefix_ << "." << primitive;
+    PrintJoinedElements("(", std::vector<typename std::common_type_t<Arg...>>{args...}, ", ", ")");
+  }
+
+  template <typename ObjType>
+  std::enable_if_t<std::is_base_of<Doc, ObjType>::value, void> PrintObject(const ObjType& obj) {
+    PrintDoc(obj);
+  }
+
+  template <typename ObjType>
+  std::enable_if_t<!std::is_base_of<Doc, ObjType>::value, void> PrintObject(const ObjType& obj) {
+    output_ << obj;
+  }
+
+  void PrintStringLiteral(const String& string) {
+    // TODO: Escape and smart quote (choose ' or " automatically)
+    output_ << "\"" << string << "\"";
+  }
+
+  void PrintNumberNode(const PrimExpr& expr) {
+    auto& dtype = expr->dtype;
+    std::ostringstream number_value;
+
+    if (expr->IsInstance<IntImmNode>()) {
+      number_value << Downcast<IntImm>(expr)->value;
+    } else if (expr->IsInstance<FloatImmNode>()) {
+      number_value.precision(17);
+      number_value << Downcast<FloatImm>(expr)->value;
+    } else {
+      LOG(FATAL) << "Do not know how to process " << expr->GetTypeKey() << " as literal number";
+    }
+
+    if (dtype == DataType::Int(32)) {
+      output_ << number_value.str();
+    } else if (dtype == DataType::Bool()) {
+      output_ << (Downcast<IntImm>(expr)->value ? "True" : "False");
+    } else {
+      PrintTIRPrimitiveCall(runtime::DLDataType2String(dtype), number_value.str());
+    }
+  };
+};
+
+class TVMScriptUnifiedPrinter;
+
+// A modified version of NodeFunctor which supports auto downcast
+class DocProducerRegistry {
+ private:
+  using InternalFType = std::function<Doc(const ObjectRef&, TVMScriptUnifiedPrinter&)>;
+
+  template <typename NodeRef, typename DocType>
+  using ProducerType = DocType (*)(const NodeRef&, TVMScriptUnifiedPrinter&);
```

Comment:
shall we try this?

```suggestion
  template <typename NodeRef, typename DocType>
  using ProducerType = std::function<DocType(const NodeRef&, TVMScriptUnifiedPrinter*));
```

Reply:
This will fail the type deduction. https://stackoverflow.com/a/60029273 I can't find a good solution to workaround this because we do need to capture the `NodeRef` type param for down casting.

Reply:
I found a good workaround by using the function_traits template from https://stackoverflow.com/a/7943765 to extract arg type from lambda. The code looks like
```
  template <typename FLambda>
  ObjectGenericFunction& register_func(FLambda func) {
    using ObjectRefType = typename function_traits<FLambda>::template arg<0>::type;
    using ObjectType = typename ObjectRefType::ContainerType;
...
```

Not sure if this is too complicated or if there is a better way